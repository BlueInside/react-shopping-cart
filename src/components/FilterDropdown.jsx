import { useEffect, useRef, useState } from 'react';
import Button from './Button';
import PropTypes from 'prop-types';
import { OptionsContainer } from './styles/OptionsContainer.styles';

function FilterDropdown({ className, setSortOption }) {
  const [isOpen, setIsOpen] = useState(false);

  const ref = useRef();

  // Closes dropdown when clicked outside component
  useEffect(() => {
    const handleClickOutside = (event) => {
      if (ref.current && !ref.current.contains(event.target)) {
        setIsOpen(false);
      }
    };

    document.addEventListener('mousedown', handleClickOutside);

    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, []);
  function toggleDropdown() {
    setIsOpen(!isOpen);
  }

  function handleDropdownItemClick(option) {
    setSortOption(option);

    toggleDropdown();
  }
  return (
    <div className={className} open={!isOpen} ref={ref} role={'dropdownContainer'}>
      <Button
        border
        role={'filter'}
        aria-label={'Sort options'}
        label={'Sort by...'}
        handleClick={toggleDropdown}
      />
      {/* {isOpen && ( */}
      <OptionsContainer open={isOpen} role="dropdownOptionsContainer">
        <Button
          border
          role={'filterOption'}
          label={'A-Z'}
          handleClick={() => handleDropdownItemClick('az')}
        />
        <Button
          border
          role={'filterOption'}
          label={'Z-A'}
          handleClick={() => handleDropdownItemClick('za')}
        />
        <Button
          border
          role={'filterOption'}
          label={'Price: Low-High'}
          handleClick={() => handleDropdownItemClick('lowHigh')}
        />
        <Button
          border
          role={'filterOption'}
          label={'Price: High-Low'}
          handleClick={() => handleDropdownItemClick('highLow')}
        />
      </OptionsContainer>
      {/* )} */}
    </div>
  );
}

FilterDropdown.propTypes = {
  setSortOption: PropTypes.func.isRequired,
  className: PropTypes.string,
};

export default FilterDropdown;
