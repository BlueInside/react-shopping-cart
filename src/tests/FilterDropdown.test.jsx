import { describe, expect, it, vi } from 'vitest';
import FilterDropdown from '../components/FilterDropdown';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { ThemeProvider } from 'styled-components';
import theme from '../components/styles/theme.';

describe('FilterDropdown component', () => {
  const customRender = (ui, options) =>
    render(<ThemeProvider theme={theme}>{ui}</ThemeProvider>, options);

  it('Renders FilterDropdown correctly', () => {
    const { container } = customRender(
      <FilterDropdown setSortOption={() => {}} />
    );
    expect(container).toMatchSnapshot();
  });
  it('Should display FilterDropdown button', () => {
    customRender(<FilterDropdown setSortOption={() => {}} />);

    const dropdownBtn = screen.getByRole('filter');
    expect(dropdownBtn).toBeInTheDocument();
  });

  it('Display FilterDropdown content on click, hide when user chose option', async () => {
    const user = userEvent.setup();
    const mockSort = vi.fn();
    customRender(<FilterDropdown setSortOption={mockSort} />);

    const dropdownBtn = screen.getByRole('filter');
    await user.click(dropdownBtn);

    const dropdownOptions = screen.getAllByRole('filterOption');
    expect(dropdownOptions).toHaveLength(4);

    await user.click(dropdownOptions[0]);
    expect(mockSort).toHaveBeenCalled();

    expect(screen.queryByRole('filterOption')).toBeNull();
  });

  it('closes the dropdown when clicked outside', async () => {
    const user = userEvent.setup();
    customRender(<FilterDropdown setSortOption={() => {}} />);

    // Open the dropdown
    user.click(screen.getByRole('filter'));

    // Click anywhere on the screen
    await user.click(document.body);

    // Check if dropdown is closed
    screen.debug(screen.queryByRole('dropdownOptionsContainer'));
    expect(screen.queryByRole('dropdownOptionsContainer')).toBeNull();
  });
});
