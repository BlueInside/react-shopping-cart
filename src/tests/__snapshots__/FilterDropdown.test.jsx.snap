// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FilterDropdown component > Renders FilterDropdown correctly 1`] = `
<div>
  <div
    role="dropdownContainer"
  >
    <button
      class="sc-beySPh goVLFM"
      role="filter"
    >
      Sort by...
    </button>
  </div>
</div>
`;
